File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/executing/AnyInCodeSystemEvaluator.java
package org.opencds.cqf.cql.engine.elm.executing;

import org.hl7.elm.r1.CodeSystemRef;
import org.opencds.cqf.cql.engine.execution.State;

public class AnyInCodeSystemEvaluator {
    public static Object internalEvaluate(Object codes, CodeSystemRef codeSystemRef, Object codeSystem, State state) {
        Object cs = null;
        if (codeSystemRef != null) {
            cs = CodeSystemRefEvaluator.toCodeSystem(codeSystemRef, state);
        } else if (codeSystem != null) {
            cs = codeSystem;
        }

        if (codes == null || cs == null) return null;

        if (codes instanceof Iterable) {
            Object result;
            for (Object code : (Iterable<?>) codes) {
                result = InCodeSystemEvaluator.inCodeSystem(code, cs, state);
                if (result instanceof Boolean && (Boolean) result) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/AnyInCodeSystemEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.hl7.elm.r1.CodeSystemRef
import org.opencds.cqf.cql.engine.execution.State
import org.opencds.cqf.cql.engine.runtime.Boolean
import org.opencds.cqf.cql.engine.runtime.List
import org.opencds.cqf.cql.engine.runtime.Value

object AnyInCodeSystemEvaluator {
    @JvmStatic
    fun internalEvaluate(
        codes: Value?,
        codeSystemRef: CodeSystemRef?,
        codeSystem: Value?,
        state: State?,
    ): Boolean? {
        var cs: Value? = null
        if (codeSystemRef != null) {
            cs = CodeSystemRefEvaluator.toCodeSystem(codeSystemRef, state)
        } else if (codeSystem != null) {
            cs = codeSystem
        }

        if (codes == null || cs == null) return null

        if (codes is List) {
            for (code in codes) {
                if (InCodeSystemEvaluator.inCodeSystem(code, cs, state)?.value == true) {
                    return Boolean.TRUE
                }
            }
        }

        return Boolean.FALSE
    }
}
